xtrack/madng_interface.py

```diff
@@ -391,11 +515,481 @@ def prepare(self, force=False):
+
+                self.optics_target_locations.add(loc)
+
+                aux = ''
+                if qty in OPTFUN_QUANTITIES:
+                    aux = 'optfun = true'
+                elif qty in ['x', 'px', 'y', 'py', 't', 'pt']:
+                    aux = f'orbit = {['x', 'px', 'y', 'py', 't', 'pt'].index(qty) + 1}'
+
+                if qty_orig == 'zeta':
+                    self._needs_zeta_scale.append(i)
+                elif qty_orig == 'delta':
+                    self._needs_delta_scale.append(i)
+
+                # set string for quantity mapping + loc to save in madng
+                targets_map_str += f"{XSUITE_MADNG_ENV_NAME}.targets_arr[{i+1}] = {{ loc = '{loc}', qty = '{qty}', {aux} }}\n"
+                xs_ng_target_map += f"{XSUITE_MADNG_ENV_NAME}.xs_ng_target_map['{qty_orig}'] = '{qty}'\n"
+
+                self.optics_target_quantities.add(qty_orig)
+
+
+            elif hasattr(target, "start") and hasattr(target, "end"):
+                loc_start = target.start if target.start != "__ele_start__" else None
+                loc_end   = target.end   if target.end   != "__ele_stop__" else None
+
+                if isinstance(target, xt.TargetRelPhaseAdvance):
+                    self.match_opt = True
+                    qty_orig = target.var
+                    qty = qty_orig[:-3] if qty_orig.endswith('_ng') else XS_NG_MAP[qty_orig]
+
+                    assert qty in TPSA_ALLOWED_TARGETS, f"Target quantity '{target.var}' not allowed with TPSA matching."
+
+                    if loc_start is not None:
+                        self.optics_target_locations.add(loc_start)
+                    if loc_end is not None:
+                        self.optics_target_locations.add(loc_end)
+                    self.optics_target_quantities.add(qty_orig)
+
+                    targets_map_str += f"{XSUITE_MADNG_ENV_NAME}.targets_arr[{i+1}] = {{ loc = '{loc_end}', qty = '{qty}', loc_start = '{loc_start}', optfun = true }}\n"
+                    xs_ng_target_map += f"{XSUITE_MADNG_ENV_NAME}.xs_ng_target_map['{target.var}'] = '{qty}'\n"
+
+                elif isinstance(target, xt.TargetRmatrixTerm):
+                    self.match_rmat = True
+                    qty = target.term
+                    tag = target.tag.split('_')[0]
+                    idx = int(tag)
+
+                    self.rmat_start_end_list[idx] = (loc_start, loc_end)
+                    self.rmat_tags.append(target.tag)
+
+                    targets_map_str += f"{XSUITE_MADNG_ENV_NAME}.targets_arr[{i+1}] = {{ loc = '{loc_end}', qty = '{qty}', loc_start = '{loc_start}', rmat = true, tag = {tag} }}\n"
+                    xs_ng_target_map += f"{XSUITE_MADNG_ENV_NAME}.xs_ng_target_map['{target.term}'] = '{qty}'\n"
+
+                else:
+                    raise NotImplementedError(f"Target of type {type(target)} not implemented for MAD-NG TPSA matching.")
+
```

Comment:
There is no `else:`, which means if someone provides a broken/wrong target, the error will pass silently. What is the expected type in l. 658? I assume `Target`, a subclass, or some other compatible ducktyped construct, in which case, maybe just change that line to `else:` and let whatever errors happen in that branch, in case the user passed a wrong thing.

There's some repeated code between the branches.

Also as discussed irl, at some point let's discuss the interface of `Target`.